src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualChar.cs

```diff
@@ -23,33 +25,53 @@ namespace Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars
     /// </summary>
     internal readonly struct VirtualChar : IEquatable<VirtualChar>
     {
-        public readonly char Char;
+        public readonly Rune Rune;
         public readonly TextSpan Span;
 
-        public VirtualChar(char @char, TextSpan span)
+        public VirtualChar(char ch, TextSpan span)
```

Comment:
Is this safe where `ch` is an incomplete surrogate? I guess `TryAddMultiCharacterEscape` passes incomplete surrogates to this constructor. For instance an emoji 😊 can be represented as three forms in C# string literals: `"😊"`, `"\U0001F60A"`, and `"\ud83d\ude0a"`. The former two are OK but `"\ud83d\ude0a"` is split into `\ud83d` and `\ude0a` in `TryAddMultiCharacterEscape` and replaced into two `\ufffd`s in `Rune`.
And more complex case is half-escaped surrogate pairs: `\ud83d\\ude0a` and `\\ud83d\ude0a`.

Reply:
i will check and validate these.  It should be the case that the caller only uses that if things were legal.  But more testing can't hurt here :)

Reply:
> And more complex case is half-escaped surrogate pairs: \ud83d\\ude0a and \\ud83d\ude0a.

What are the difference between these two?  I'm not seeing it.

Reply:
Nevermind :)  i see what you're saying.  one is escaped, one is the actual char.  Will validate that.

Reply:
Great cases.  You def caught a bug here.  Thanks so much! :)

Reply:
@ufcpp Any other interesting cases you would like me to consider? i think the algorithm is correct now, but it can't hurt to check more corners.

Reply:
No. As far as I know, problematic case is only incomplete surrogates.
FYI, the C# lexer/parser have the same problem. I have been trying to make C# support non-BMP  characters in identifiers ([working branch](https://github.com/ufcpp/roslyn/commits/surrogate-pair-identifier)). Some characters including 𓄿 (U+1313F, Egyptian hieroglyph), 𒀀 (U+12000, Cuneiform), 𩸽 (U+29E3D, minor Chinese character) and so on are not BMP but letter characters which should be able to be used in identifiers. I'm struggling with these corners, especially "one is escaped, one is the actual char" cases.

Reply:
In so far as identifiers, there are more complex corner cases but it maybe doesn't matter to `VirtualChar`.

[C# spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#identifiers) says "Any formatting_characters are removed" in identifiers. So...

- `ab` and `a\u200Db` are considered the same
- `\uD867\uDE3D` and `\uD867\u200D\uDE3D` should be considered the same as `𩸽`

Reply:
I think half-escaped case is ignorable because it is invalid UTF-8 sequence. However, `\uD867\uDE3D` and `\uD867\u200D\uDE3D` are valid.

Reply:
One solution is to make C# prohibit escaped incomplete surrogates explicitly by spec. But this is a breaking change.

Reply:
> C# spec says "Any formatting_characters are removed" in identifiers. So...

Reply:
> One solution is to make C# prohibit escaped incomplete surrogates explicitly by spec. But this is a breaking change.

Fortunately, i'm not trying to change the language here.  Just expose a useful abstraction for letting people know what characters are in a string-literal for IDE purposes.  For our cases the requirements are only:

1. we shouldn't crash.
2. for legal code, we shouldn't claim it is something different than what it is.

What we are allowed to then do is:

1. not bother doing anything on broken code.
2. for legal but strange code (like half-surrogates), we can just not offer the features.  These should be so rare as to not matter for our purposes.